Mathlib/Topology/MetricSpace/Lipschitz.lean

```diff
@@ -581,6 +590,127 @@ end Metric
 
 end LipschitzOnWith
 
+namespace LocallyLipschitz
+variable [PseudoEMetricSpace α] [PseudoEMetricSpace β] [PseudoEMetricSpace γ] {f : α → β}
+
+/-- A Lipschitz function is locally Lipschitz. -/
+protected lemma _root_.LipschitzWith.locallyLipschitz {K : ℝ≥0} (hf : LipschitzWith K f) :
+    LocallyLipschitz f :=
+  fun _ ↦ ⟨K, univ, Filter.univ_mem, lipschitzOn_univ.mpr hf⟩
+
+/-- The identity function is locally Lipschitz. -/
+protected lemma id : LocallyLipschitz (@id α) := LipschitzWith.id.locallyLipschitz
+
+/-- Constant functions are locally Lipschitz. -/
+protected lemma const (b : β) : LocallyLipschitz (fun _ : α ↦ b) :=
+  (LipschitzWith.const b).locallyLipschitz
+
+/-- A locally Lipschitz function is continuous. (The converse is false: for example,
+$x ↦ \sqrt{x}$ is continuous, but not locally Lipschitz at 0.) -/
+protected theorem continuous {f : α → β} (hf : LocallyLipschitz f) : Continuous f := by
+  apply continuous_iff_continuousAt.mpr
+  intro x
+  rcases (hf x) with ⟨K, t, ht, hK⟩
+  exact (hK.continuousOn).continuousAt ht
+
+/-- The composition of locally Lipschitz functions is locally Lipschitz. --/
+protected lemma comp  {f : β → γ} {g : α → β}
+    (hf : LocallyLipschitz f) (hg : LocallyLipschitz g) : LocallyLipschitz (f ∘ g) := by
+  intro x
+  -- g is Lipschitz on t ∋ x, f is Lipschitz on u ∋ g(x)
+  rcases hg x with ⟨Kg, t, ht, hgL⟩
+  rcases hf (g x) with ⟨Kf, u, hu, hfL⟩
+  -- idea: shrink t to ensure it is mapped to u
+  -- more precisely: restrict g to t' := t ∩ g⁻¹(u); the preimage of u under g':=g∣t.
+  let g' := t.restrict g
+  set t' : Set α := ↑(g' ⁻¹' u) with ht'
+  have h₁ : t' = t ∩ g ⁻¹' u := by
+    rw [ht']
+    ext1 y
+    simp [Lean.Internal.coeM]
+    aesop
+  have h₂ : t' ∈ 𝓝 x := by -- FIXME: surely, there is a nicer proof
+    -- by ht, t contains an open subset U
+    rcases (mem_nhds_iff.mp ht) with ⟨U, hUt, hUopen, hxU⟩
+    -- similarly, u contains an open subset V
+    rcases (mem_nhds_iff.mp hu) with ⟨V, hVt, hVopen, hgxV⟩
+    -- by continuity, g⁻¹(u) contains the open subset g⁻¹(V)
+    have : ContinuousOn g U := (hgL.mono hUt).continuousOn
+    have h : IsOpen (U ∩ (g ⁻¹' V)) := this.preimage_open_of_open hUopen hVopen
+    have : U ∩ (g ⁻¹' V) ⊆ t' := by rw [h₁]; apply inter_subset_inter hUt (preimage_mono hVt)
+    -- now, U ∩ g⁻¹(V) is an open subset contained in t'
+    rw [mem_nhds_iff]
+    use U ∩ (g ⁻¹' V)
+    exact ⟨this, ⟨h, ⟨hxU, hgxV⟩⟩⟩
+  have : g '' t' ⊆ u := by calc g '' t'
+    _ = g '' (t ∩ g ⁻¹' u) := by rw [h₁]
+    _ ⊆ g '' t ∩ g '' (g ⁻¹' u) := by apply image_inter_subset
+    _ ⊆ g '' t ∩ u := by gcongr; apply image_preimage_subset
+    _ ⊆ u := inter_subset_right _ _
+  use Kf * Kg, t'
+  exact ⟨h₂, hfL.comp (hgL.mono coe_subset) (mapsTo'.mpr this)⟩
```

Comment:
```suggestion
  refine ⟨Kf * Kg, t ∩ g⁻¹' u, inter_mem ht (hg.continuous.continuousAt hu), ?_⟩ 
  exact hfL.comp (hgL.mono (inter_subset_left _ _))
    ((mapsTo_preimage g u).mono_left (inter_subset_right _ _))
```
We have essentially already everything in the API, it's just a matter of finding the right blocks and putting them together.